pravega/pravega
# upgrade hadoopVersion but workaround HADOOP-16765

**Change log description**  
The hadoopVersion is updated to 3.1.1 in this change without downgrading Guava for most of the code. It is only downgraded for test so that test may pass. This was required for the issue described by [Hadoop-16765](https://issues.apache.org/jira/browse/HADOOP-16765).  Upgrading Hadoop to the latest also required code changed which is described by the issues [HDFS-752](https://issues.apache.org/jira/browse/HDFS-752) and [HDFS-7477](https://issues.apache.org/jira/browse/HDFS-7477). The links for these issues are mentioned here so that the patch associated with these issues can be viewed. They are the only break-in changes we have. This update to HDFS mitigates the security vulnerabilities cited in the GitHub issue [4874](https://github.com/pravega/pravega/issues/4874) linked to this request.  

**Purpose of the change**  
The purpose of this change is to remove the security vulnerabilities associated with the hadoop version. The Hadoop Version poses a bigger security threat than others to Pravega.

Fixes #4874.

**What the code does**  
This code updates hadoop, downgrades guava only for test and makes the changes required for the updated hadoopVersion.

**How to verify it**  
This change was verified by travis build.

The exception faced with upgrading Guava is
```
io.pravega.storage.hdfs.HDFSStorageTest > testDelete FAILED
    java.lang.NoSuchMethodError: com.google.common.util.concurrent.Futures.addCallback(Lcom/google/common/util/concurrent/ListenableFuture;Lcom/google/common/util/concurrent/FutureCallback;)V
        at org.apache.hadoop.hdfs.server.datanode.checker.ThrottledAsyncChecker.addResultCachingCallback(ThrottledAsyncChecker.java:167)
        at org.apache.hadoop.hdfs.server.datanode.checker.ThrottledAsyncChecker.schedule(ThrottledAsyncChecker.java:156)
        at org.apache.hadoop.hdfs.server.datanode.checker.StorageLocationChecker.check(StorageLocationChecker.java:164)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.makeInstance(DataNode.java:2762)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.instantiateDataNode(DataNode.java:2677)
        at org.apache.hadoop.hdfs.MiniDFSCluster.startDataNodes(MiniDFSCluster.java:1663)
        at org.apache.hadoop.hdfs.MiniDFSCluster.initMiniDFSCluster(MiniDFSCluster.java:905)
        at org.apache.hadoop.hdfs.MiniDFSCluster.<init>(MiniDFSCluster.java:517)
        at org.apache.hadoop.hdfs.MiniDFSCluster$Builder.build(MiniDFSCluster.java:476)
        at io.pravega.storage.hdfs.HDFSClusterHelpers.createMiniDFSCluster(HDFSClusterHelpers.java:32)
        at io.pravega.storage.hdfs.HDFSStorageTest.setUp(HDFSStorageTest.java:68)
``` 
This method was removed from this library as seen by [others](https://issues.jenkins-ci.org/browse/JENKINS-56252) as well. This [issue](https://issues.apache.org/jira/browse/BEAM-3156) is tracked on the originator side as well with this [jira](https://issues.apache.org/jira/browse/HADOOP-16765). Notice the method takes two parameters instead of the [three](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/util/concurrent/FutureCallback.html) in recent versions. Note that the exception described above is being addressed with different jars where the newer one is different from the older one) and the guava blogs also calls out this issue:
https://mvnrepository.com/artifact/com.google.guava/listenablefuture
[Downgrading](https://blog.gradle.org/version-downgrade) is the only option left at this time. Fortunately, this is only required for test.
[GuavaVersion29versus25.txt](https://github.com/pravega/pravega/files/4815622/GuavaVersion29versus25.txt)

This PR differs from an earlier PR for the same GitHub issue by not downgrading Guava for production code.
